packages/block-editor/src/store/effects.js

```diff
@@ -84,12 +84,31 @@ export default {
 		const blockB = getBlock( state, clientIdB );
 		const blockBType = getBlockType( blockB.name );
 		const { clientId, attributeKey, offset } = getSelectionStart( state );
-		const hasTextSelection = (
+		const selectedBlockType = clientId === clientIdA ? blockAType : blockBType;
+		const attributeDefinition = selectedBlockType.attributes[ attributeKey ];
+		const canRestoreTextSelection = (
 			( clientId === clientIdA || clientId === clientIdB ) &&
 			attributeKey !== undefined &&
-			offset !== undefined
+			offset !== undefined &&
+			// We cannot restore text selection if the RichText identifier
+			// is not a defined block attribute key. This can be the case if the
+			// fallback intance ID is used to store selection (and no RichText
+			// identifier is set), or when the identifier is wrong.
+			attributeDefinition
 		);
 
+		if ( ! attributeDefinition ) {
+			if ( typeof attributeKey === 'number' ) {
+				window.console.error(
```

Comment:
Again this is another minor comment (which I don't think is a blocker for the PR). This seems like a very specific error, and I wonder whether it could be more generic by catching anything where value is defined but the type isn't a string, and adding something like this as the error:
```
`RichText needs an identifier prop that is the block attribute key of the attribute it controls. Its type is expected to be a string, but was ${ typeof attributeKey }`
```